### PR TITLE
Improved validation of tools during workflow execution.

### PR DESCRIPTION
**Pull request by jmchilton**
Do a more complete validation of everything after workflow parameters have been replaced and inputs connected up.

Attempt 2 at the idea in https://github.com/galaxyproject/galaxy/pull/1284.

Lets see if the tests go any better now.

### CODE REVIEW

**Comment by jmchilton** [comment]
6 failures instead of 44. Making progress.

**Comment by jmchilton** [comment]
Ugh - none of these fail for me locally. Odd, maybe because I rebased against dev before testing - forced push and we will try again.

**Comment by jmchilton** [comment]
Probably a problem with `validate` on postgres for these tools - maybe a unicode vs str thing? If that is what it is though I'm annoyed nothing would show up in the logs so I pushed out 41590e3 which should result in job creation problems from halting workflow execution (not job execute/tool failure problems, job creation problems). It also adds a lot more logging on these workflow problems. Hopefully this helps.

**Comment by jmchilton** [comment]
Progress:

```
galaxy.tools.execute: WARNING: There was a failure executing a job for tool [cat_list] - {'input1': "Value no longer valid for 'Concatenate Dataset', replaced with default"}
galaxy.tools.execute: DEBUG: Executed all jobs for tool request: (42.009 ms)
galaxy.workflow.run: ERROR: Failed to schedule Workflow[id=9name=Workflow (imported from API)], problem occurred on WorkflowStep[index=4,type=tool].
Traceback (most recent call last):
  File "/galaxy/lib/galaxy/workflow/run.py", line 155, in invoke
    jobs = self._invoke_step( step )
  File "/galaxy/lib/galaxy/workflow/run.py", line 221, in _invoke_step
    jobs = step.module.execute( self.trans, self.progress, self.workflow_invocation, step )
  File "/galaxy/lib/galaxy/workflow/modules.py", line 875, in execute
    raise Exception("Failed to create %d jobs for workflow." % count)
Exception: Failed to create 1 jobs for workflow.
```

Will fix the logging typos and log the actual reason the validation is failing.

**Comment by jmchilton** [comment]
With extra logging tests discovered a fairly substantial and unfortunate bug, fixed with 918ce64.

File: lib/galaxy/workflow/modules.py [hunk 1]
```diff
@@ -870,6 +870,12 @@ def callback( input, value, prefixed_name, prefixed_label ):
         jobs = execution_tracker.successful_jobs
         for job in jobs:
             self._handle_post_job_actions( step, job, invocation.replacement_dict )
+        if execution_tracker.execution_errors:
+            failed_count = len(execution_tracker.execution_errors)
+            success_count = len(execution_tracker.successful_jobs)
+            all_count = failed_count + success_count
+            message = "Failed to create %d out of %s job(s) for workflow step." % (failed_count, all_count)
+            raise Exception(message)
```
Comment (bgruening) [inline lib/galaxy/workflow/modules.py:886]: Should we log the exception and raise it afterwards, like in the run.py or the other way around?

**Comment by guerler** [comment]
:+1:

**Pull request merged**